guardian/typerighter
# Add and edit dictionary rules in rule manager

## What does this change?

This adds the ability to add and edit dictionary rules in the rule manager.

The rule edit panel will include fewer fields than the panel for other rules, only including the following fields: Rule type, Pattern, Source, Tags.

The Source (category) will be automatically set to 'Collins dictionary addition' when adding a new dictionary rule. This is so we can distinguish between 'Collins Dictionary' rules and rules created by Guardian users.

The Source field will be disabled for dictionary rules, to prevent users from changing this to a different non-dictionary related source. When editing existing 'Collins Dictionary' rules, it will no longer be possible for users to change the source.

Previously the logic was using the rule category to determine which rules to include when instantiating the Dictionary matcher. This PR changes this so that a single dictionary matcher is created for all rules with the 'dictionary' rule type.  This is to enable us to assign different categories/sources to newly created dictionary rules.

There are outstanding questions on whether it should be possible for users to batch edit dictionary rules at the same time as different rule types. The logic added in this PR that defaults dictionary rules to have a specific rule source (when creating/editing a single rule) could have an impact on batch editing. There are also some preexisting quirks with batch editing that could have an impact on batch editing different rule types - I captured these issues in a [Trello card](https://trello.com/c/pFQUahTm/1770-address-batch-editing-issues).

Trello card is [here](https://trello.com/c/u5JXhu6y/1701-enable-create-edit-and-unpublish-archive-of-dictionary-rules-in-the-manager).

## How to test

- run the Rule Manager & Composer locally

Adding a rule:
- start creating a dictionary rule
- select rule type 'dictionary' - does the source automatically get set to 'Guardian dictionary addition'?
- can you change the rule source? (you shouldn't be able to)

Editing a rule
- edit an existing dictionary rule
- test making changes, discarding changes, etc.
- click 'Unpublish Rule' - this should turn the rule from a Live into a Draft rule
- test whether the rule word you unpublished is now appearing with a red underline in Composer (this can take a bit of time to update)

## Images

### Before
![Screenshot 2023-10-11 at 17 31 39](https://github.com/guardian/typerighter/assets/7883129/bd57bd2d-e528-411a-b66d-1b8444ea43e9)

### After
![Screenshot 2023-10-11 at 17 34 08](https://github.com/guardian/typerighter/assets/7883129/a3150563-369b-4551-ab95-278434421bde) - ![Screenshot 2023-10-11 at 17 15 05](https://github.com/guardian/typerighter/assets/7883129/cb4e7acd-ad27-42da-bc0b-72c16bb4412f)
